### DISCUSSION TITLE
Removing “hidden colours” from ImageMagick's histogram output

### DISCUSSION THREAD

**Discussion by ghost** [ImageMagick/ImageMagick#2884] [Help] [answered]
Hi, I have a PNG image below that has 3 strips of colour - #72C7FF, #D9FF7A, and #FF7A7A. No other colours are present to the naked eye and the background is 100% transparent.

![1](https://user-images.githubusercontent.com/68956224/100623639-54372b00-331a-11eb-96a5-2940cf6e6e79.png)

However, ImageMagick's “histogram” output is showing coloured pixels that don't appear present on the image. This is the command I'm using:

`convert 1.png -alpha disassociate -format %c histogram:info: | awk '{print $1, $3}' | sort -V -r | head -n 10`

Here's the output I'm getting (it shows the top 10 coloured pixels):

> 42385: #000000
> 38682: #FF7A7A
> 38682: #72C7FF
> 38679: #D9FF7A
> 927: #FFFFFF
> 33: #FFFF93
> 33: #FF9393
> 33: #8AF0FF
> 28: #FF8686
> 28: #FF8383

#72C7FF, #D9FF7A, and #FF7A7A all appear towards the top as expected but where are all those extra colours coming from? In particular, the 42385 counts of #000000 pixels and 927 counts of #FFFFFF? How can I configure the command above to remove any “hidden colours“ that don't appear to the naked eye?

Thank you!

Edit: I'm using ImageMagick 6.9.11-5 on macOS Mojave 10.14.6

**Comment by snibgo** [comment]
What version of IM, on what platform?

> ...and the background is 100% transparent.

No, it isn't. If you do a "magick in.png -verbose info:" you will see many values of alpha. Most pixels are either fully transparent or fully opaque, but some are semi-transparent.

The output from your "convert" includes alpha values, but your "awk" ignores them.

**Reply by ghost** [reply]
ImageMagick 6.9.11-5, with macOS Mojave version 10.14.6 and Terminal Version 2.9.5

**Reply by ghost** [reply]
Thanks for the info @Snibgo. Unless you have better ideas, I'll try then and see if I can create a loop or something to only output the pixel if it contains a “1” at the end of its srbga value.

**Comment by fmw42** [comment]
Try -alpha off rather than -alpha disassociate.  

Or if your alpha channel is binary

convert image sparse-color: | tr " " "\n"

See https://imagemagick.org/script/formats.php

**Reply by ghost** [reply]
Thanks `-alpha off` was a step in the right direction. It removed the histogram output of:

> 927: #FFFFFF

Now it's just to figure out how to handle the output of `42385: #000000`

Right now I'm working with this:

`convert in.png -background '#FF69B4' -alpha remove -alpha off -format '%c' histogram:info: | awk '{print $1, $3}' | sort -V -r | head -n 10`

That at least removes the transparent pixels, replacing it with a background color of #FF69B4 (HotPink), and then maybe I can loop through the histogram's output to weed it out…

Although perhaps there's an in-built ImageMagick process for this, I'm yet to find it if it exists.

**Comment by ghost** [comment, marked as answer]
Ok, if anyone is interested, I've found my acceptable answer. Here's my command.

`convert in.png -background '#FF69B4' -alpha remove -alpha off -format '%c' histogram:info: | awk '{print $1, $3}' | sed '/#FF69B4/d' | sort -V -r | head -n 10`

I added sed to just remove the line containing the fully transparent pixel I didn't want in my histogram output.

I'll leave this question unanswered for the time being to see if it gets any further suggestions to further complete this answer (i.e. remove all the semi-transparent pixels too or use a more in-built ImageMagick process instead to achieve the same/better result.)

**Comment by fmw42** [comment]
Do you really need both -alpha remove and -alpha off? I would think not. One or the other should be sufficient.

**Reply by ghost** [reply]
I would've thought so too but adding in the extra `-alpha off` shortens the hex colors to 6 digits instead of 8 (which is what I need.) And if I don't include the additional -`alpha remove` the black pixels aren't converted to #FF69B4 (HotPink).

**Reply by fmw42** [reply]
Yes, you are correct. I had to review the alpha methods at https://imagemagick.org/Usage/masking/#alpha. Your command should be equivalent, I believe, to 

`-background '#FF69B4' -alpha background -alpha off`
